File: src/main/java/minic/uilocal/panel/MiniCBottomPanelModelFactory.java
package minic.uilocal;

import minic.uiapi.UiDiagnosticDto;
import minic.uiapi.UiGlobalDataDto;
import minic.uiapi.UiRealtimeAnalysisDto;
import minic.uiapi.UiStageDataDto;

import java.util.List;

/**
 * 根据 UI API DTO 生成底部面板内容。
 */
public final class MiniCBottomPanelModelFactory {
    /**
     * 创建底部面板模型。
     *
     * @param stageData 当前阶段数据
     * @param globalData 全局数据
     * @return 底部面板模型
     */
    public MiniCBottomPanelModel create(UiStageDataDto stageData, UiGlobalDataDto globalData) {
        return create(stageData, globalData, null);
    }

    /**
     * 创建底部面板模型。
     *
     * @param stageData 当前阶段数据
     * @param globalData 全局数据
     * @param realtimeAnalysis 实时分析数据
     * @return 底部面板模型
     */
    public MiniCBottomPanelModel create(
            UiStageDataDto stageData,
            UiGlobalDataDto globalData,
            UiRealtimeAnalysisDto realtimeAnalysis
    ) {
        List<String> problems = problems(stageData, globalData, realtimeAnalysis);
        List<String> output = stageData == null
                ? List.of("等待开始观测会话")
                : stageData.accumulatedOutput().isEmpty() ? List.of(stageData.currentItem()) : stageData.accumulatedOutput();
        List<String> terminal = terminal(stageData);
        return new MiniCBottomPanelModel(problems, output, terminal);
    }

    private List<String> problems(
            UiStageDataDto stageData,
            UiGlobalDataDto globalData,
            UiRealtimeAnalysisDto realtimeAnalysis
    ) {
        if (realtimeAnalysis != null) {
            if (realtimeAnalysis.diagnostics().isEmpty()) {
                return List.of("OK  实时分析通过");
            }
            return realtimeAnalysis.diagnostics().stream()
                    .map(diagnostic -> diagnostic.severity() + "  " + diagnostic.code() + "  " + diagnostic.message())
                    .toList();
        }
        List<UiDiagnosticDto> diagnostics;
        if (globalData != null && !globalData.diagnostics().isEmpty()) {
            diagnostics = globalData.diagnostics();
        } else if (stageData != null && !stageData.diagnostics().isEmpty()) {
            diagnostics = stageData.diagnostics();
        } else {
            return List.of("OK  暂无 diagnostics");
        }
        return diagnostics.stream()
                .map(diagnostic -> diagnostic.severity() + "  " + diagnostic.code() + "  " + diagnostic.message())
                .toList();
    }

    private List<String> terminal(UiStageDataDto stageData) {
        if (stageData == null) {
            return List.of("PS> minic observe <source.mc>");
        }
        return List.of(
                "PS> minic observe --stage " + stageData.stage(),
                stageData.stage() + "[" + stageData.completedSteps() + "/" + stageData.totalSteps() + "] " + stageData.currentItem()
        );
    }
}

File: uiweb/src/panel/MiniCBottomPanelModelFactory.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import type {
  UiDiagnosticDto,
  UiGlobalDataDto,
  UiRealtimeAnalysisDto,
  UiStageDataDto,
} from "../translation/uiapi";
import { MiniCBottomPanelModel } from "./MiniCBottomPanelModel";

export const miniCBottomPanelModelFactoryMirror = {
  "javaPath": "src/main/java/minic/uilocal/panel/MiniCBottomPanelModelFactory.java",
  "webPath": "uiweb/src/panel/MiniCBottomPanelModelFactory.ts",
  "packageName": "minic.uilocal",
  "exportName": "MiniCBottomPanelModelFactory",
  "kind": "class",
  "imports": [
    "java.util.List",
    "minic.uiapi.UiDiagnosticDto",
    "minic.uiapi.UiGlobalDataDto",
    "minic.uiapi.UiRealtimeAnalysisDto",
    "minic.uiapi.UiStageDataDto"
  ],
  "fields": [],
  "methods": [
    {
      "name": "create",
      "signature": "create(UiStageDataDto stageData,UiGlobalDataDto globalData,UiRealtimeAnalysisDto realtimeAnalysis)"
    },
    {
      "name": "create",
      "signature": "create(UiStageDataDto stageData,UiGlobalDataDto globalData)"
    },
    {
      "name": "problems",
      "signature": "problems(UiStageDataDto stageData,UiGlobalDataDto globalData,UiRealtimeAnalysisDto realtimeAnalysis)"
    },
    {
      "name": "terminal",
      "signature": "terminal(UiStageDataDto stageData)"
    }
  ]
} as const satisfies JavaMirrorFile;

export class MiniCBottomPanelModelFactory {
  static readonly mirror = miniCBottomPanelModelFactoryMirror;

  readonly mirror = miniCBottomPanelModelFactoryMirror;

  create(
    stageData: UiStageDataDto | null,
    globalData: UiGlobalDataDto | null,
    realtimeAnalysis: UiRealtimeAnalysisDto | null = null,
  ): MiniCBottomPanelModel {
    const problems = this.problems(stageData, globalData, realtimeAnalysis);
    const output =
      stageData === null
        ? ["等待开始观测会话"]
        : stageData.accumulatedOutput.length === 0
          ? [stageData.currentItem]
          : stageData.accumulatedOutput;
    const terminal = this.terminal(stageData);
    return new MiniCBottomPanelModel(problems, output, terminal);
  }

  problems(
    stageData: UiStageDataDto | null,
    globalData: UiGlobalDataDto | null,
    realtimeAnalysis: UiRealtimeAnalysisDto | null,
  ): readonly string[] {
    if (realtimeAnalysis !== null) {
      if (realtimeAnalysis.diagnostics.length === 0) {
        return ["OK  实时分析通过"];
      }
      return realtimeAnalysis.diagnostics.map(formatDiagnostic);
    }
    const diagnostics =
      globalData !== null && globalData.diagnostics.length > 0
        ? globalData.diagnostics
        : stageData !== null && stageData.diagnostics.length > 0
          ? stageData.diagnostics
          : null;
    if (diagnostics === null) {
      return ["OK  暂无 diagnostics"];
    }
    return diagnostics.map(formatDiagnostic);
  }

  terminal(stageData: UiStageDataDto | null): readonly string[] {
    if (stageData === null) {
      return ["PS> minic observe <source.mc>"];
    }
    return [
      `PS> minic observe --stage ${stageData.stage}`,
      `${stageData.stage}[${stageData.completedSteps}/${stageData.totalSteps}] ${stageData.currentItem}`,
    ];
  }

  summary(): string {
    return `MiniCBottomPanelModelFactory: ${this.mirror.methods.length} methods, ${this.mirror.fields.length} fields`;
  }
}

function formatDiagnostic(diagnostic: UiDiagnosticDto): string {
  return `${diagnostic.severity}  ${diagnostic.code}  ${diagnostic.message}`;
}

export default MiniCBottomPanelModelFactory;
